### ISSUE TITLE
[Gameplay] Movement Points on First Turn

### ISSUE THREAD

**Issue by JoLiKMC** [ihhub/fheroes2#6316]
* Split from #6310

### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

In doing some testing, I've noticed that in the original game, heroes will **always start with a set amount of Movement Points** – specifically, the Warlock hero **starts with 12(00) movement points** and the Wizard hero **starts with 11** while every _other_ hero **starts with 10**.  It doesn't matter if a hero starts with only Ultra Fast units (Champion, Phoenix), has any-or-all movement-enhancing Artifacts equipped (Nomad Boots, Traveler's Boots, True Compass), or has _any_ level of Logistics, they will **always** start with a set amount of Movement Points on their first turn.  This is **absolutely not** the case with _fheroes2_. wherein Heroes seem to start with the "correct" amount of Movement Points on Day 1, based on their starting inventory / skills / army.

Again, I'm not sure if this is a bug in the original or something, but I guess _someone_ noticed it… and "fixed" it.

### Save file

[HoMM2 Map - Hero Movement Test.zip](https://github.com/ihhub/fheroes2/files/10167201/HoMM2.Map.-.Hero.Movement.Test.zip)

**Comment by JoLiKMC** [comment]
@oleg-derevenetz commented [18 hours ago](https://github.com/ihhub/fheroes2/issues/6310#issuecomment-1339573371):
> Regarding this one - there was a discussion at the time and this considered as a bug in the original game - because there is absolutely no way to entirely remove an army from a hero using the regular game UI. If this was a regular feature, then the developers of the original game would certainly have made some way for this. But neither in this nor in other main games of the series (HoMM3, HOMMV) there is no such functionality.

**Comment by oleg-derevenetz** [comment]
> This is absolutely not the case with fheroes2. wherein Heroes seem to start with the "correct" amount of Movement Points on Day 1, based on their starting inventory / skills / army.

Then I don't see how this is a bug. If this is a bug, it's a bug in the original game.

**Comment by JoLiKMC** [comment]
> Then I don't see how this is a bug. If this is a bug, it's a bug in the original game.

I did a little more research into the matter.  This behavior, in vanilla _Heroes II_, seems **intentional**:
* In _Heroes I_, Barbarians and Warlocks get **12(00) Movement Points** on their first turn.  Knights and Sorceresses get 10.
**Every turn after**, Movement Points are based on a hero's slowest army unit and inventory.
* In _Heroes II_, Warlocks get **12 Movement Points** and Wizards get **11 Movement Points** on their first turn.  Everyone else gets 10.
**Every turn after**, Movement Points are based on a hero's slowest army unit, secondary skills, buffs, and inventory.
* In _Heroes III_, everyone gets movement points based on their army, secondary skills, buffs, and inventory **on all turns**.

So basically, whether or not this is a "bug" in _fheroes2_ depends on what direction y'all are trying to take this project.

* If you're trying to make the game as authentic to _Heroes II_ as possible, then **this is a bug**.
* If you're trying to make the game play more like _Heroes III_ (and future titles?), then **this is not a bug**.

**Comment by Branikolog** [comment]
Hello, @JoLiKMC !
I agree with @oleg-derevenetz . It's not intuitive, when at first turn you have a certain amount of movement points, but on the second turn it is changed. The most obvious and predictable behaviour in this situation is to allow hero to have the amount of movement points he **should** have, no matter what day it is.

You mentioned before:

> wherein Heroes seem to start with the "correct" amount of Movement Points on Day 1

So even you feel the current implementation as "correct", while the original behaviour is tricky and not obvious to new players.

**Comment by LeHerosInconnu** [comment]
Hello @JoLiKMC, @oleg-derevenetz, and @Branikolog,

I think it's a bug from the original game (similar to the hero's full movement bug after a retreat or a reddition), and fheroes2 should fix this.
Army troops, secondary skills and artifacts must be taken into account from the start of the scenario (and not only at the start of the second turn).

In the original game, the starting default hero has a movement according to the troops actually present in his army (in the example a Barbarian with and without an orc troop by default).

In original game:

![Hero start move 001](https://user-images.githubusercontent.com/43583994/206243614-fa91044f-eebb-43e4-91a8-432f6af7ed39.png)

![Hero start move 002](https://user-images.githubusercontent.com/43583994/206243632-f831978c-eb38-4837-9159-208d77deb613.png)

**Comment by JoLiKMC** [comment]
I… think you nailed it on the head, @LeHerosInconnu.  The Day 1 movement points are generated from **the Standard troops** of a class _regardless_ of whether they have said units in their army or not.  Some class heroes start with only **Level 1 units**, but most classes starts with both **Level 1 and Level 2units**.  I was _absolutely not_ taking the Level 2 units into account.  And in testing this in _Heroes I_, again, that's exactly how it works.  _It's based on each individual hero's starting army._

Everything makes _sense_, now!  In _Heroes II_, the Wizard's slowest starting creature would normally be the Halfling (Level 1, Slow speed), they get the Slow penalty (-4 movement points), but the Necromancer's slowest starting creature is the Zombie (Level 2, Very Slow speed), they start with the Very Slow penalty (-5 movement points).  However, if a Necromancer hero starts with _only_ Skeletons (Level 1, Average speed), they'll only have the Average movement penalty (-3 movement points)!  Similar things are at play with the Sorceress and Barbarian, too.

From this, I can determine that there is no "class starting movement points".  There never was.  Rather, it's determined by the specific hero's Standard troops.  And I mean that _very_ literally.  If a hero _normally_ starts with some Peasants and some Archers (Level 1 and 2, both Very Slow), but you make them start with just a pile of Phoenixes or Champions (the fastest units in the game), the hero will _still_ have the Very Slow movement penalty on Day 1.  It's almost like the game loads a hero with their Standard troops, sets the Movement Points accordingly, and _then_ gives the hero their custom units, Artifacts, and Secondary Skills.  In fact… that may well be what's going on, here.  The original _Heroes I_ and _II_ are very likely setting Movement Points **as the very first thing** before a given turn.  And it's this very simple oversight that leads to the wonky movement points on Day 1.

Man.  It took a lot of comments to get here, but I finally understand.  I'm finally ready to admit that, yes, this was an **oversight** in the original _Heroes of Might and Magic I / II_.  Therefore, the way _fheroes2_ does it – not to mention the way _Heroes of Might and Magic III_ (and future titles) do it – is the correct, _intended_ way.  _All of that_ being said… I finally agree that this **should not be reverted** in _fheroes2_.

Thank you, LeHerosInconnu.  I don't know how I missed something so obvious.  But I mean, so did New World Computing and CyberLore.  _Twice_.  Hah.  And thank you, everyone else, for putting up with this.  Sometimes, it takes a village.  Ya know?
